File: _legacy-swift/SundeeFundee/Domain/CycleProgramGenerator.swift
import Foundation

// MARK: - CycleProgramGenerator

/// Adapts a Program's exercises for the user's current cycle phase.
/// Returns the same program unchanged when adaptation is disabled or no phase is available.
enum CycleProgramGenerator {

    static func adaptProgram(
        _ program: Program,
        phase: CyclePhase?,
        settings: CycleSettings?,
        preferences: CycleAdaptationPreferences?,
        periodLogs: [PeriodLog],
        referenceDate: Date = .now,
        readinessScore: Double? = nil
    ) -> Program {
        // Guest users or opt-out: return unchanged
        guard let preferences, preferences.adaptationEnabled else { return program }
        guard let cycleSettings = settings else { return program }

        let policy = CycleAdaptationPolicy()

        // Determine effective phase
        let status = CycleCalculations.calculateCycleStatus(
            periodLogs: periodLogs,
            settings: cycleSettings,
            referenceDate: referenceDate
        )
        let currentPhase  = status?.currentPhase
        let effectivePhase = policy.resolvePhase(
            currentPhase: currentPhase,
            lastKnownPhase: phase,
            fallbackPhase: preferences.fallbackPhase
        )

        // Determine confidence
        let lastPeriodStart = periodLogs.max(by: { $0.startDate < $1.startDate })?.startDate
        let confidence = policy.resolveConfidence(
            currentPhase: currentPhase,
            lastKnownPhase: phase,
            periodLogCount: periodLogs.count,
            lastPeriodStart: lastPeriodStart,
            referenceDate: referenceDate
        )

        let readiness = policy.resolveReadinessTier(readinessScore: readinessScore)

        // Adapt each week's sessions
        let adaptedWeeks = program.weeks.map { week in
            ProgramWeek(
                week: week.week,
                phaseID: week.phaseID,
                isTestWeek: week.isTestWeek,
                sessions: week.sessions.map { session in
                    ProgramSession(
                        sessionID: session.sessionID,
                        sessionName: session.sessionName,
                        sessionType: session.sessionType,
                        focus: session.focus,
                        exercises: session.exercises.map { ex in
                            let adapted = policy.applyPhaseAdjustment(
                                exercise: ex,
                                phase: effectivePhase,
                                readinessTier: readiness,
                                confidence: confidence,
                                profile: program.cycleAdjustmentProfile
                            )
                            let cycleNote = Self.cycleAdjustmentNote(
                                original: ex,
                                adapted: adapted,
                                phase: effectivePhase
                            )
                            guard let cycleNote else { return adapted }
                            let combinedNote = [ex.notes, cycleNote]
                                .compactMap { $0 }
                                .joined(separator: " · ")
                            return ProgramExercise(
                                exercise: adapted.exercise,
                                variant: adapted.variant,
                                sets: adapted.sets,
                                reps: adapted.reps,
                                percent1RM: adapted.percent1RM,
                                restMinutes: adapted.restMinutes,
                                notes: combinedNote,
                                bodyweightOnly: adapted.bodyweightOnly
                            )
                        }
                    )
                }
            )
        }

        return Program(
            id: program.id,
            name: program.name,
            category: program.category,
            description: program.description,
            durationWeeks: program.durationWeeks,
            sessionsPerWeek: program.sessionsPerWeek,
            difficulty: program.difficulty,
            phases: program.phases,
            cycleAdjustmentProfile: program.cycleAdjustmentProfile,
            weeks: adaptedWeeks
        )
    }

    // MARK: - Note generation

    /// Returns a note string describing the original prescription and the cycle-adjusted values,
    /// or nil if no meaningful adjustment was made.
    static func cycleAdjustmentNote(
        original: ProgramExercise,
        adapted: ProgramExercise,
        phase: CyclePhase
    ) -> String? {
        let origSets = original.sets.description
        let origReps = original.reps.description
        let adaptSets = adapted.sets.description
        let adaptReps = adapted.reps.description

        let origLoad: String? = original.percent1RM.map { "\(Int(($0 * 100).rounded()))%" }
        let adaptLoad: String? = adapted.percent1RM.map { "\(Int(($0 * 100).rounded()))%" }

        let setsChanged = origSets != adaptSets
        let repsChanged = origReps != adaptReps
        let loadChanged = origLoad != adaptLoad

        guard setsChanged || repsChanged || loadChanged else { return nil }

        let origScheme = formatScheme(sets: origSets, reps: origReps, load: origLoad)
        let adaptScheme = formatScheme(sets: adaptSets, reps: adaptReps, load: adaptLoad)

        return "Program: \(origScheme) → \(phase.displayName): \(adaptScheme)"
    }

    private static func formatScheme(sets: String, reps: String, load: String?) -> String {
        if let load {
            return "\(sets)×\(reps) @ \(load)"
        }
        return "\(sets)×\(reps)"
    }
}

File: src/domain/cycle/cycle-program-generator.ts
// =============================================================================
// CycleProgramGenerator — TypeScript port of CycleProgramGenerator.swift
// Adapts a Program's exercises for the user's current cycle phase.
// Pure function — returns a new Program object, never mutates input.
// =============================================================================

import { calculateCycleStatus } from './cycle-calculations';
import { applyPhaseAdjustment, resolveReadinessTier } from './cycle-adaptation-policy';
import type { CyclePhase, CycleSettings, PeriodLog, Program } from '../types';
import type { AdaptationConfidence } from './cycle-adaptation-policy';

// ---------------------------------------------------------------------------
// Confidence resolution (mirrors Swift CycleAdaptationPolicy.resolveConfidence)
// ---------------------------------------------------------------------------

/** @internal Exported for testing only */
export function resolveConfidence(
  currentPhase: CyclePhase | undefined,
  periodLogCount: number,
  lastPeriodStart: Date | undefined,
  referenceDate: Date
): AdaptationConfidence {
  if (periodLogCount === 0) return 'low';
  if (lastPeriodStart !== undefined) {
    const daysSince = (referenceDate.getTime() - lastPeriodStart.getTime()) / 86400000;
    if (daysSince > 45) return 'low';
  }
  if (currentPhase !== undefined && periodLogCount >= 3) return 'high';
  if (currentPhase !== undefined) return 'medium';
  return 'low';
}

// ---------------------------------------------------------------------------
// Main adapt function
// ---------------------------------------------------------------------------

/**
 * Adapt all sessions in a program for the user's current cycle phase.
 *
 * Mirrors CycleProgramGenerator.adaptProgram(_:phase:settings:preferences:periodLogs:referenceDate:readinessScore:).
 *
 * If periodLogs is empty, calculateCycleStatus returns undefined and the
 * function falls back to 'follicular' with low confidence. The program is
 * still adapted (unlike Swift which has an adaptationEnabled guard — the RN
 * adaptation is always applied since the caller controls whether to invoke it).
 *
 * @param program - The program to adapt
 * @param periodLogs - User's period log history
 * @param settings - User's cycle settings
 * @param referenceDate - Reference date (defaults to today)
 * @param readinessScore - Optional readiness survey score (1–10)
 */
export function adaptProgram(
  program: Program,
  periodLogs: PeriodLog[],
  settings: CycleSettings,
  referenceDate: Date = new Date(),
  readinessScore?: number
): Program {
  const status = calculateCycleStatus(periodLogs, settings, referenceDate);
  const currentPhase = status?.currentPhase;

  // Fallback to follicular when no cycle data is available
  const effectivePhase: CyclePhase = currentPhase ?? 'follicular';

  const lastPeriodStart = periodLogs.length > 0
    ? periodLogs.reduce((latest, log) =>
        log.startDate > latest ? log.startDate : latest,
        periodLogs[0].startDate
      )
    : undefined;

  const confidence = resolveConfidence(
    currentPhase,
    periodLogs.length,
    lastPeriodStart,
    referenceDate
  );

  const readinessTier = resolveReadinessTier(readinessScore);

  const adaptedSessions = program.sessions.map(session => ({
    ...session,
    exercises: session.exercises.map(exercise =>
      applyPhaseAdjustment(exercise, effectivePhase, readinessTier, confidence)
    ),
  }));

  return {
    ...program,
    sessions: adaptedSessions,
  };
}
